feat(mme): amf client servicer layer

panyogesh opened the pull request:
Fixes:
  1. Removed the proto from AMF cmake as thats handled in oai/lib
  2. Introduced amf client servicer which will abstratct amf's client
     interaction like subscriberdb
  3. Introduced the Mock layer in amf unit test for the client testing
  4. Fixed the CMake for unit testing of amf

Testing:
  1. Basis unit testing
  2. Basic snaity with UERANSIM

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

<!-- Enumerate changes you made and why you made them -->

## Test Plan

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

panyogesh committed:
feat(mme): amf client servicer layer

Fixes:
  1. Removed the proto from AMF cmake as thats handled in oai/lib
  2. Introduced amf client servicer which will abstratct amf's client
     interaction like subscriberdb
  3. Introduced the Mock layer in amf unit test for the client testing
  4. Fixed the CMake for unit testing of amf

Testing:
  1. Basis unit testing
  2. Basic snaity with UERANSIM

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/include/amf_client_servicer.h:
```diff
@@ -0,0 +1,57 @@
+      const void* resync_info, uint8_t resync_info_len, amf_ue_ngap_id_t ue_id);
+};
+
+// Initilaze the client servicer layer
```
typo `Initilaze`

panyogesh replied:
Comments addressed.

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_client_servicer.cpp:
```diff
@@ -0,0 +1,57 @@
+  auto authentication_client =
+      std::make_shared<AsyncM5GAuthenticationServiceClient>();
+
+  amf_client_servicer_g = AmfClientServicer(authentication_client);
```
what is the significance of `_g`? is it to signify that this is grpc client? it does not help much to readability.

ulaskozat replied:
ah this is to indicate that it is global?

panyogesh replied:
Yes..intention to denote it as global...This layer will serve as interface to all clients. After the merge it will be extended t o serve mobilityd and sessiond clients.

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_client_servicer.cpp:
```diff
@@ -0,0 +1,57 @@
+
+namespace magma5g {
+
+AmfClientServicer amf_client_servicer_g;
```
The overall pattern looks more complex than it needs to be. You are wrapping a singleton class with another class and then instantiating a global instance. Instead of this, it makes more sense to convert singleton class into a non singleton one and define a global. Then you can write you mock class directly for that.

ulaskozat commented:
@panyogesh reading your new comment and I understand that you want this class as an entry class for all the grpc clients. In this case, do we still need to define actual GRPC clients as singletons?

ulaskozat commented:
@panyogesh I think the better way is to avoid the global definition here completely. Let's create a struct that is a holder of each of the globals here (i.e., it is a meta class) and pass the pointer to that into `argp` parameter (currently passing as NULL) in `itti_create_task`. 

We can then mock each of the clients individually for mobilityd, sessiond, etc.

panyogesh committed:
feat(mme): amf client servicer layer

Fix:
  1. Introduced a amf metadata struct.
  2. Metadata struct stores the client related servicers
     and will be passed on to the amf thread.
  3. In other places insted of calling individual clients
     the wrapper will be called and its also used for Unit test
     extensions.

Test:
  1. Basic sanity with Simulators
  2. Run the unit tests.

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

panyogesh commented:
> @panyogesh I think the better way is to avoid the global definition here completely. Let's create a struct that is a holder of each of the globals here (i.e., it is a meta class) and pass the pointer to that into `argp` parameter (currently passing as NULL) in `itti_create_task`.
> 
> We can then mock each of the clients individually for mobilityd, sessiond, etc.

1. Created the global struct : amf_metadata
2. Initialized during startup before launching the amf thread and it stores the information about clients (in this patch it stores authenticaiton client)
3. Its then passed on to the amf_thread.

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_app_main.cpp:
```diff
@@ -155,7 +156,9 @@ static int handle_message(zloop_t* loop, zsock_t* reader, void* arg) {
   itti_mark_task_ready(TASK_AMF_APP);
-  const task_id_t tasks[] = {TASK_NGAP, TASK_SERVICE303};
+  const task_id_t tasks[]      = {TASK_NGAP, TASK_SERVICE303};
+  amf_metadata_t* amf_metadata = (amf_metadata_t*) args;
```
please use `reinterpret_cast`.

panyogesh replied:
Comments addressed.

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_app_main.cpp:
```diff
@@ -182,7 +185,11 @@ extern "C" int amf_app_init(const amf_config_t* amf_config_p) {
   create_state_matrix();
-  if (itti_create_task(TASK_AMF_APP, &amf_app_thread, NULL) < 0) {
+
+  // Initialze the metadata and client service entries
```
Typo `Initialze`

panyogesh replied:
Comments addressed.

panyogesh committed:
feat(mme): Amf Client Service layer

Review comments addressed.

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/include/amf_metadata_util.h:
```diff
@@ -0,0 +1,44 @@
+
+} amf_metadata_t;
+
+void amf_metadata_intialize(amf_metadata_t* metadata_p);
```
```suggestion
void amf_metadata_initialize(amf_metadata_t* metadata_p);
```

panyogesh replied:
Comments addressed.

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/include/amf_metadata_util.h:
```diff
@@ -0,0 +1,44 @@
+ * This strucuture is used in itti_create_task
+ * Container of amf structure/class at the task level.
+ */
+extern amf_metadata_t amf_metadata;
```
What's the constraint of this being exposed as global? I think from the AMF app's thread scope this should be dynamically allocated (on the thread initialization procedure) This way this global definition can be avoided. Also let's make sure if that's being allocated the memory is being freed when the task context is destroyed on the thread exit function. cc @ulaskozat

ulaskozat replied:
+1

panyogesh replied:
Comments addressed.

panyogesh committed:
feat(mme): Addressing review comments.

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

ardzoht reviewed (approved):
lgtm

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_metadata_util.cpp:
```diff
@@ -0,0 +1,30 @@
+#include "include/amf_client_servicer.h"
+#include <memory>
+
+std::shared_ptr<amf_metadata_t> metadata_info_ptr{};
+
+/* Function to initialize amf metadata information */
+void amf_metadata_initialize(
+    const std::shared_ptr<amf_metadata_t>& metadata_p) {
+  metadata_info_ptr = std::move(metadata_p);
+
+  metadata_info_ptr->amf_client_servicer_init();
+}
+
+std::shared_ptr<magma5g::AmfClientServicer> amf_get_client_servicer_ref() {
+  return metadata_info_ptr->amf_client_servicer;
+}
```
The original comment was about to eliminate globals. The pattern should be like the following:
1. Create meta data object during task initialization.
2. Have new function `init_task_context_with_args`  (in the same file as `init_task_context`) and pass the meta data  pointer as the arguments for handler function. Inside `init_task_context_with_args`, pass the argument to zloop_reader.
3. In handler function, pass the pointer for metadata to called functions if they need to access the meta data.

cc @ardzoht

panyogesh replied:
Tried making the change. Metadata parameter will be used by function : start_authentication_information_procedure().  This function is coming at a stack depth of 16. So the changes are spanning across many files/functions.  Need to discuss on the same.

ardzoht reviewed (changes requested):
Let's remove use of global shared ptr and update this per @ulaskozat' comment

panyogesh commented:
Closing as the comments as its addressed by #9618
